File: c_code/room_generating.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct Passage
{
    int *pattern;
    void *from;
    void *to;
} Passage;

typedef struct SqRoom
{
    int *pattern;
    char directions[4];
    Passage **passages;
} SqRoom;

// сверху структуры, снизу функции

int *RandomPattern() // твой random в питоне
{
    int *pattern = (int *)malloc(sizeof(int));
    *pattern = rand(); // random value for the pattern
    return pattern;
};

Passage **InitializePassage() // просто выделяем память для passage
{
    Passage **new_passage = (Passage **)malloc(4 * sizeof(Passage *));
    for (int i = 0; i < 4; i++)
    {
        new_passage[i] = (Passage *)malloc(sizeof(Passage));
        new_passage[i]->pattern = RandomPattern();
        new_passage[i]->from = NULL;
        new_passage[i]->to = NULL;
    }
    return new_passage;
}

SqRoom *InitializeRoom() // просто выделяем память для sqroom
{
    SqRoom *new_room = (SqRoom *)malloc(sizeof(SqRoom));
    new_room->pattern = RandomPattern();
    new_room->directions[0] = 'r';
    new_room->directions[1] = 'l';
    new_room->directions[2] = 'u';
    new_room->directions[3] = 'd';
    new_room->passages = InitializePassage();
    return new_room;
}

void CheckNeighbors(SqRoom ****graph, int index_j, int index_i, SqRoom **current_room)
{
    for (int current_direction = 0; current_direction < 4; ++current_direction)
    {
        SqRoom *neighbor = NULL;
        switch (current_direction)
        {
        case 0:                  // Right
            if (index_i + 1 < 4) // Assuming 4 is the grid size
                neighbor = (*graph)[index_j][index_i + 1];
            break;
        case 1: // Left
            if (index_i - 1 >= 0)
                neighbor = (*graph)[index_j][index_i - 1];
            break;
        case 2: // Up
            if (index_j - 1 >= 0)
                neighbor = (*graph)[index_j - 1][index_i];
            break;
        case 3:                  // Down
            if (index_j + 1 < 4) // Assuming 4 is the grid size
                neighbor = (*graph)[index_j + 1][index_i];
            break;
        }

        if (neighbor != NULL)
        {
            (*current_room)->passages[current_direction]->from = *current_room;
            (*current_room)->passages[current_direction]->to = neighbor;
            free(neighbor);
        }
    }
}

void CreateMap(bool **existing_table, struct SqRoom ***graph, int index_i, int index_j)
{
    if (graph[index_j][index_i])
    {
        return;
    }
    else if (!graph[index_j][index_i] && existing_table[index_j][index_i] == true)
    {
        graph[index_j][index_i] = InitializeRoom();
        CheckNeighbors(&graph, index_j, index_i, &graph[index_j][index_i]);

        if (index_i + 1 < 4)
            CreateMap(existing_table, graph, index_i + 1, index_j); // Right
        if (index_i - 1 >= 0)
            CreateMap(existing_table, graph, index_i - 1, index_j); // Left
        if (index_j - 1 >= 0)
            CreateMap(existing_table, graph, index_i, index_j - 1); // Up
        if (index_j + 1 < 4)
            CreateMap(existing_table, graph, index_i, index_j + 1); // Down
    }
    else
    {
        return;
    }
}

int main()
{
    printf("hello\n");
    return 0;
}

File: c_code/py_version/room_generating.py
import random

class Passage:
    def __init__(self, pattern=None, from_room=None, to_room=None):
        self.pattern = pattern if pattern is not None else self.random_pattern()
        self.from_room = from_room
        self.to_room = to_room

    @staticmethod
    def random_pattern():
        return random.randint(0, 100)  # Generate random pattern


class SqRoom:
    def __init__(self):
        self.pattern = Passage.random_pattern()
        self.passages = {
            'r': Passage(),
            'l': Passage(),
            'u': Passage(),
            'd': Passage()
        }


def check_neighbors(graph, index_j, index_i, current_room):
    directions = ['r', 'l', 'u', 'd']
    neighbors_coords = [(0, 1), (0, -1), (-1, 0), (1, 0)]

    for direction, (dj, di) in zip(directions, neighbors_coords):
        neighbor_j = index_j + dj
        neighbor_i = index_i + di

        if 0 <= neighbor_j < len(graph) and 0 <= neighbor_i < len(graph[0]):
            neighbor = graph[neighbor_j][neighbor_i]

            if neighbor:
                current_room.passages[direction].from_room = current_room
                current_room.passages[direction].to_room = neighbor


def create_map(existing_table, graph, index_i, index_j):
    if graph[index_j][index_i]:
        return
    elif not graph[index_j][index_i] and existing_table[index_j][index_i]:
        graph[index_j][index_i] = SqRoom()
        check_neighbors(graph, index_j, index_i, graph[index_j][index_i])

        if index_i + 1 < len(graph[0]):
            create_map(existing_table, graph, index_i + 1, index_j)  # Right
        if index_i - 1 >= 0:
            create_map(existing_table, graph, index_i - 1, index_j)  # Left
        if index_j - 1 >= 0:
            create_map(existing_table, graph, index_i, index_j - 1)  # Up
        if index_j + 1 < len(graph):
            create_map(existing_table, graph, index_i, index_j + 1)  # Down


# Example usage
if __name__ == "__main__":
    size = 4
    existing_table = [[False for _ in range(size)] for _ in range(size)]
    graph = [[None for _ in range(size)] for _ in range(size)]

    # Simulate existing table with some rooms
    existing_table[1][1] = True
    existing_table[1][2] = True
    existing_table[2][2] = True

    create_map(existing_table, graph, 1, 1)
    
    # Display the generated map (for testing purposes)
    for row in graph:
        for room in row:
            if room:
                print("Room", room.pattern, "Passages", {k: v.to_room.pattern if v.to_room else None for k, v in room.passages.items()})
            else:
                print("None", end=" ")
        print()
